layer/contracts/runs.py

```diff
@@ -61,41 +60,41 @@ def local_relative_paths(self) -> Iterator[str]:
 )
 
 
-class FunctionDefinition(abc.ABC):
+class FunctionDefinition:
     def __init__(
         self,
-        func: Any,
+        func: Callable[..., Any],
         project_name: str,
         account_name: str,
+        asset_type: AssetType,
+        asset_name: str,
+        fabric: Fabric,
+        asset_dependencies: List[AssetPath],
+        pip_dependencies: List[str],
+        resource_paths: List[ResourcePath],
+        assertions: List[Assertion],
         version_id: Optional[uuid.UUID] = None,
         repository_id: Optional[uuid.UUID] = None,
         description: str = "",
         uri: str = "",
-        language_version: Tuple[int, int, int] = _language_version(),
     ) -> None:
         self.func = func
         self.func_name: str = func.__name__
+
         self.project_name = project_name
         self.account_name = account_name
+        self.asset_type = asset_type
+        self.asset_name = asset_name
+        self.fabric = fabric
+        self.asset_dependencies = asset_dependencies
+        self.pip_dependencies = pip_dependencies
+        self.resource_paths = resource_paths
+        self.assertions = assertions
+
         self.version_id = version_id
         self.repository_id = repository_id
         self.description = description
         self.uri = uri
-        self.language_version = language_version
-
-        layer_settings = func.layer
-        name = layer_settings.get_asset_name()
-        if name is None:
-            raise LayerClientException("Name cannot be empty")
-        self.name = name
-        self.resource_paths = layer_settings.get_resource_paths()
-        fabric = layer_settings.get_fabric()
-        if fabric is None:
-            fabric = Fabric.default()
-        self._fabric = fabric
-        self._dependencies: List[AssetPath] = layer_settings.get_dependencies()
-        self.pip_packages = layer_settings.get_pip_packages()
-        self.pip_requirements_file = layer_settings.get_pip_requirements_file()
 
         self.func_source = inspect.getsource(self.func)
 
```

Comment:
so, the `source_code_digest`, also `uri` above etc are all dataset-specific. I guess one downside of this "bundling" is that models will have these unnecessarily as well..

Reply:
yes, true, but with several more factors we should be able to get datasets and models closer and more consistent with each other. for example, no reason why `source_code_digest` and `uri` can't be attributes of both datasets and models.